### ISSUE TITLE
Packaging questions

### ISSUE THREAD

**Issue by gllmhyt** [gkarsay/parlatype#6]
I made a quick&dirty PKGBUILD for Arch Linux, I'm open to suggestions: https://aur.archlinux.org/packages/parlatype/

```PKGBUILD
pkgname=parlatype
pkgver=1.5
pkgrel=1
pkgdesc="GNOME audio player for transcription"
arch=('any')
url="https://gkarsay.github.io/parlatype/"
license=('GPL')
depends=('intltool' 'gobject-introspection-runtime' 'yelp-tools' 'gst-plugins-base' 'gst-plugins-good')
source=("https://github.com/gkarsay/$pkgname/releases/download/v$pkgver/$pkgname-$pkgver.tar.gz")
noextract=()
md5sums=('8de6ae7bb0f261b4f48effbe133684c5')

prepare() {
	cd "$pkgname-$pkgver"
}

build() {
	cd "$pkgname-$pkgver"
	autoreconf
	./configure --prefix=/usr
	make
}

check() {
	cd "$pkgname-$pkgver"
	make -k check
}

package() {
	cd "$pkgname-$pkgver"
	make DESTDIR="$pkgdir/" install
}
```

**Comment by ferdnyc** [comment]
Along those lines, here's the RPM spec file I used to package Parlatype on Fedora 25, based heavily off of the parlatype-daily Ubuntu recipe in Launchpad. It's _definitely_ a work-in-progress version and would never pass Fedora packaging review, as it flagrantly violates the guidelines in several ways.

(For starters...)
* I should really be packaging libparlatype as a separate RPM, but I'm not.
* As a result I had to call `%find_lang` with the `--all-name` argument so it would pick up the `libparlatype.mo` locale files.
* Mucking with the `$RPM_BUILD_ROOT` tree to move the Libreoffice helper file(s) from `/usr/lib/` to `/usr/lib64/`, _after_ `%make_install`, is a huge no-no; I really should be configuring with `--without-libreoffice-helpers` and manually installing the helper file as suggested in the README.

Still, it does produce seemingly-working `parlatype-1.5` and `parlatype-libreoffice-helpers-1.5` RPMs.

```specfile
Name:           parlatype
Version:        1.5
Release:        0.1%{?dist}
Summary:        GNOME audio player for transcription

License:        GPLv3+
URL:            http://gkarsay.github.io/parlatype/
Source0:        https://github.com/gkarsay/parlatype/releases/download/v%{version}/%{name}-%{version}.tar.gz

#Patch0:         parlatype-revert-abf2d5e.patch

BuildRequires:  intltool >= 0.50.2
BuildRequires:  gtk3-devel >= 3.10.0
BuildRequires:  gtk-doc
BuildRequires:  glade-devel >= 3.15.0
BuildRequires:  gobject-introspection
BuildRequires:  gstreamer1-devel
BuildRequires:  gstreamer1-plugins-base-devel
BuildRequires:  libxml2
BuildRequires:  perl-File-FcntlLock

Requires:       gstreamer1-plugins-good%{?_isa}

%description
Parlatype is a minimal audio player for manual speech transcription, written for the GNOME desktop environment. It plays audio sources to transcribe them in your favourite text application.

%package        devel
Summary:        Development files for %{name}
Requires:       %{name}%{?_isa} = %{version}-%{release}

%description    devel
The %{name}-devel package contains libraries and header files for
developing applications that use %{name}.

%package        libreoffice-helpers
Summary:        Tools for integrating %{name} with LibreOffice
Requires:       %{name}%{?_isa} = %{version}-%{release}

%description    libreoffice-helpers
The %{name}-devel package contains script files for LibreOffice to better integrate with %{name}.


%prep
%setup -q
#%patch0 -p1 -b .style

%build
%configure --disable-static --enable-gtk-doc --enable-glade-catalog
%make_build


%install
rm -rf $RPM_BUILD_ROOT
%make_install
find $RPM_BUILD_ROOT -name '*.la' -exec rm -f {} ';'

# Libreoffice scripts should be installed in arch libdir
%ifarch x86_64
mv $RPM_BUILD_ROOT/usr/lib/libreoffice $RPM_BUILD_ROOT/%{_libdir}/
%endif

%find_lang %{name} --all-name --with-gnome

%check
appstream-util validate-relax --nonet $RPM_BUILD_ROOT/%{_datadir}/appdata/org.gnome.parlatype.appdata.xml
desktop-file-validate $RPM_BUILD_ROOT/%{_datadir}/applications/org.gnome.parlatype.desktop


%post -p /sbin/ldconfig

%postun -p /sbin/ldconfig


%files -f %{name}.lang
%license COPYING
%doc README.md INSTALL
%{_bindir}/%{name}
%{_libdir}/libparlatype.so.*
%{_libdir}/girepository-1.0/Parlatype-1.0.typelib
%{_datadir}/appdata/org.gnome.parlatype.appdata.xml
%{_datadir}/applications/org.gnome.parlatype.desktop
%{_datadir}/glib-2.0/schemas/org.gnome.parlatype.gschema.xml
%{_datadir}/icons/hicolor/*/apps/%{name}.png
%{_mandir}/man1/%{name}.1*

%files devel
%doc README.md INSTALL
%{_includedir}/*
%{_libdir}/libparlatype.so
%{_libdir}/pkgconfig/%{name}.pc
%{_datadir}/glade/catalogs/%{name}.xml
%{_datadir}/glade/pixmaps/*
%{_datadir}/gir-1.0/Parlatype-1.0.gir
%{_datadir}/gtk-doc/*

%files libreoffice-helpers
%{_libdir}/libreoffice/share/Scripts/python/*.py*

%changelog
* Mon Jun 26 2017 FeRD (Frank Dana) <ferdnyc AT gmail com> - 1.5-0.1
- Initial build
```

**Comment by gkarsay** [comment]
Thank you for your efforts distributing Parlatype! I'm not familiar with Arch Linux and Fedora packaging, I can just give a few guesses what might be missing.

* For the PKGBUILD: The depends list seems to be too short. There's a list (Debian-based, however) in the README. Add everything that's not covered by base-devel. It think Gtk is missing. In Debian I split it into several packages, the important ones for users are the library, the program/application itself and the libreoffice-helpers. Users don't need the introspection data and the documentation. You could configure with `--disable-introspection`. Maybe you have to change the libreoffice path as for Fedora, I don't know.

* The RPM looks good. perl-File-FcntlLock in BuildRequires suppresses some warnings during the build in Debian, not sure you need it, too. The description of the libreoffice-helpers package seems to have the wrong name. I can't comment much more on that.

* PKGBUILD and RPM: A missing dependency for the libreoffice-helpers is libreoffice-script-provider-python (that's the package name in Debian), I have to add it to the README. Maybe it's included it in your LibreOffice package, then it should depend on LibreOffice I guess.

Since this is not a bug report, I will close the issue sooner or later as not-a-bug. I'm leaving it open for now for discussion.

**Comment by ferdnyc** [comment]
> A missing dependency for the libreoffice-helpers is libreoffice-script-provider-python (that's the package name in Debian), I have to add it to the README. Maybe it's included it in your LibreOffice package, then it should depend on LibreOffice I guess.

Yeah, I was puzzling over _exactly_ which package that would be, and never came to a solid conclusion so I punted for the moment (since I knew full well my packaging of the helper script was broken and dirty regardless). We don't have a package as clearly-named as -script-provider-python, alas. The possible choices, in Fedora's packaging, would seem to be:
```
libreoffice-ure-common.noarch : Common UNO Runtime Environment
libreoffice-ure.x86_64 : UNO Runtime Environment
libreoffice-ure.i686 : UNO Runtime Environment
libreoffice-pyuno.x86_64 : Python support for LibreOffice
```
Most _likely_ it's the last one. (Made all the _more_ likely by the fact that it requires `libreoffice-ure(x86-64)`, which in turn requires `libreoffice-ure-common`, so requiring `libreoffice-pyuno(x86-64)` would suffice to chain the whole stack in.) So, I'll just pop that in there. ...Done.

There's still the issue that parlatype's `libreoffice/Makefile.am` _really_ needs to use `$(libdir)` instead of `$(prefix)` to specify its install path — unless someone can think of a better, more portable method for looking up libreoffice installed paths — because at least in Fedora's packaging, `libreoffice/share/` resides in the appropriate arch-dependent `/usr/lib/` or `/usr/lib64/`.

**Comment by gkarsay** [comment]
I changed the install dir for libreoffice-helpers to $(libdir) in commit 73e7085bcbeaa341ed0736a3dcfd99938db9d1ac. Hope this helps.

**Comment by ferdnyc** [comment]
Broke my build, until I removed the ugly hack of a workaround from the `.spec` file, but I'm glad to see the back of it. Much cleaner now.

**Comment by gkarsay** [comment]
I made a change in commit ea8010bc8db09b2b7c98d633b08f11474ded5ee8 that affects the RPM package: The names of some of the installed files changed as I also changed the application ID. Everything that had org.gnome.parlatype in the name is now com.github.gkarsay.parlatype, that affect 5 file names in the spec file.

**Comment by ferdnyc** [comment]
Got it, thanks. Spec updated accordingly.

**Comment by postblue** [comment]
I updated the file so it makes the difference between build-needed and run-needed libraries:

```PKGBUILD
# Maintainer: Guillaume Hayot <ghayot@postblue.info>
pkgname=parlatype
pkgver=1.5
pkgrel=4
pkgdesc="GNOME audio player for transcription"
arch=('any')
url="https://gkarsay.github.io/parlatype/"
license=('GPL')
depends=('gtk3' 'gst-plugins-base' 'gst-plugins-good')
makedepends=('gobject-introspection' 'libreoffice' 'yelp-tools' 'intltool' 'glade' 'gtk-doc')
optdepends==('libreoffice: LibreOffice macros')
source=("https://github.com/gkarsay/$pkgname/releases/download/v$pkgver/$pkgname-$pkgver.tar.gz")
noextract=()
md5sums=('8de6ae7bb0f261b4f48effbe133684c5')

prepare() {
	cd "$pkgname-$pkgver"
}

build() {
	cd "$pkgname-$pkgver"
	autoreconf
	# Parlatype ships its own library, libparlatype. Developers might be interested in having a library documentation,
	# gobject introspection and a glade catalog for the widgets. These are the configure options:
	# --enable-gtk-doc: install library documentation (default: no)
	# --enable-introspection: install gobject introspection (default: yes)
	# --enable-glade-catalog: install a glade catalog (default: no)
	# --with-libreoffice: install LibreOffice macros (default: yes)
	# If you want the program only, you would use --prefix=/usr --disable-introspection.
	./configure --prefix=/usr --enable-gtk-doc --enable-glade-catalog
	make
}

check() {
	cd "$pkgname-$pkgver"
	make -k check
}

package() {
	cd "$pkgname-$pkgver"
	make DESTDIR="$pkgdir/" install
}
```
I hope this is fine!

**Comment by gkarsay** [comment]
I've decided to keep this "issue" open. In case you have a working package for your distro, could you please leave some instructions here how to use it, how to install it. I'd like to put them on the [homepage](http://gkarsay.github.io/parlatype/#installation).

**Comment by gkarsay** [comment]
There is a **new stable version 1.5.3** available. The tarball ships also some unit tests but they will only work in a runtime environment (with display) and will probably fail in a pure build environment, therefore I recommend to configure with `--disable-tests` for packaging.

@gllmhyt: Is your aur package working? I guess command line instructions would be `pacman -S parlatype`?

**Comment by gllmhyt** [comment]
@gkarsay: I updated [my PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=parlatype) and it's working on my laptop (I still need to configure a minimal chroot to test dependecies, but so far so good).

**Comment by gkarsay** [comment]
Thank you, btw. I had to release version 1.5.4 because of a regression on older GTK+ versions, so for a rolling release there is no need to update to 1.5.4.

**Comment by gllmhyt** [comment]
@gkarsay Still updated as I try not to forget releases.

**Comment by gkarsay** [comment]
There is a **new stable version 1.5.5 of Parlatype** available. The tarball can be found on the release page: https://github.com/gkarsay/parlatype/releases/tag/v1.5.5

There should be no changes needed for packaging this version.

**Comment by gllmhyt** [comment]
Thanks! I updated `pkgver` in the PKGBUILD and sent it right away to AUR.

**Comment by gkarsay** [comment]
There is a new minor stable version 1.5.6 of Parlatype available. The tarball can be found on the release page: https://github.com/gkarsay/parlatype/releases/tag/v1.5.6

There should be no changes needed for packaging this version.

**Comment by ferdnyc** [comment]
Hmm. Parlatype seems like a natural fit for Fedora, but I see that it's currently still not included in the dist repos. Because it _can_ be built with MP3 support, it _may_ be eligible for [rpmfusion](http://rpmfusion.org/) (where I am a packager). Although now that the MP3 patents have expired, MP3 support may no longer make a package ineligible for inclusion in Fedora (where I am not a packager).

I'll put together a 1.5.6 package spec that actually conforms to the guidelines (not like that slapdash attempt I made previously), and see if I can't get that process started, wherever it needs to happen. Even if it does need to go into Fedora, it's probably time I got myself established as a Packager there, too. 

(Ultimately we'd want to get a version into Fedora either way, even if it needs to be built without the MP3 support. In that case, rpmfusion would host `parlatype-freeworld` as an alternative with MP3 support included. That's the current arrangement for Audacity and similar.)

With any luck, we'll be able to see Parlatype included in Fedora 29 when it releases in October / November.

**Comment by ferdnyc** [comment]
> Parlatype uses GStreamer for audio playback. There is no switch to build it with or without MP3 support.

Yeah, I'd slightly misinterpreted this part of `README.md`:

> Required runtime dependencies:
> 
> * GTK+ 3
> * GStreamer
> * GStreamer "Good" Plugins
>
> Optional runtime dependencies, to support MP3 files:
> 
> * GStreamer "Ugly" Plugins

..._**However**_, most likely due to the recent MP3 patent expirations, `libgstmpg123.so` is now part of `gstreamer-plugins-GOOD`, and is included even in Fedora's official package repo. So, I'm not sure `gstreamer-plugins-ugly` is really required at all for MP3, even as a runtime dependency. 

(It would be if you wanted to load audio from MPEG-2 files, as `libgstmpeg2dec.so` is in there. And I'm guessing the AAC audio codec for MP4 audio as well, though I'm having trouble locating the actual plugin file for that one.)

**Comment by ferdnyc** [comment]
Regardless, it's now clear to me that Parlatype belongs in Fedora proper, so I'll need to get started on the process of making that happen.

**Comment by gkarsay** [comment]
I'm closing this issue. If you have other particular questions about Parlatype packaging, feel free to open a new issue. I will inform the known packagers about new releases, probably around March 2019.
